src/main/java/org/folio/circulation/services/LogCheckInService.java

```diff
@@ -0,0 +1,65 @@
+package org.folio.circulation.services;
+
+import static org.folio.circulation.support.Result.succeeded;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.folio.circulation.domain.CheckInLogRecord;
+import org.folio.circulation.domain.CheckInProcessRecords;
+import org.folio.circulation.storage.CheckInStorageRepository;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.ClockManager;
+import org.folio.circulation.support.HttpFailure;
+import org.folio.circulation.support.Result;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LogCheckInService {
+  private static final Logger log = LoggerFactory.getLogger(LogCheckInService.class);
+
+  private final CheckInStorageRepository checkInStorageRepository;
+
+  public LogCheckInService(Clients clients) {
+    this.checkInStorageRepository = new CheckInStorageRepository(clients);
+  }
+
+  public CompletableFuture<Result<CheckInProcessRecords>> logCheckInOperation(
+    CheckInProcessRecords checkInProcessRecords) {
+
+    CompletableFuture.runAsync(() -> {
+      final CheckInLogRecord checkInLogRecord = CheckInLogRecord.builder()
+        .withOccurredDateTime(ClockManager.getClockManager().getDateTime())
+        .withItemId(checkInProcessRecords.getItem().getItemId())
+        .withCheckInServicePointId(checkInProcessRecords.getCheckInServicePointId().toString())
+        .withPerformedByUserId(checkInProcessRecords.getLoggedInUserId())
+        .build();
+
+      logCheckInOperation(checkInLogRecord);
+    });
+
+    return CompletableFuture.completedFuture(succeeded(checkInProcessRecords));
+  }
+
+  private void logCheckInOperation(CheckInLogRecord checkInLogRecord) {
+    checkInStorageRepository.createCheckInLogRecord(checkInLogRecord)
+      .handle((result, ex) -> {
+        if (result.failed() || ex != null) {
+          logCheckInFailure(result.cause(), ex);
+        } else {
+          log.debug("Check-in operation recorded successfully");
+        }
+
+        return null;
+      });
+  }
```

Comment:
Is it intended that this operation can fail silently and let the rest of the process continue?

Reply:
Yes, the process won't block the check-in flow and will be executed in a separate thread. This approach has been chosen in order to not slow down the main check-in flow. 

If a failure occurred, it will be logged in `WARN` level with the `Result.cause().getReason()` details.

Reply:
I think this decision needs reviewing with Emma. Personally, I think the whole process should fail if the fact that a check in has occurred cannot be recorded.

I think that is important, because if we intend to start modelling this process with more autonomous sub-processes e.g. send notices, returning a loaned item, request fulfilment etc, then a critical starting point for that will be knowing a check in has occurred.

Does that make sense?

Reply:
> Yes, the process won't block the check-in flow and will be executed in a separate thread. This approach has been chosen in order to not slow down the main check-in flow.

I think this decision needs reviewing with Emma. Personally, I think the whole process should fail if the fact that a check in has occurred cannot be recorded.

I think that is important, because if we intend to start modelling this process with more autonomous sub-processes e.g. send notices, returning a loaned item, request fulfilment etc, then a critical starting point for that will be knowing a check in has occurred.

> If a failure occurred, it will be logged in `WARN` level with the `Result.cause().getReason()` details.

What concerns me about this process is that any failures may not be noticed for a long time, meaning a potential loss of data.

(I believe something similar happened for Chalmers when notice processing was failing for a while without folks noticing it)

Reply:
Yes, such kind of processes requires supervision.

I thought, there should be a centralized logging collection for production (ELK, or similar).

Reply:
It guess that most likely if `check-in-storage` will be failing then entire circulation-storage will be failing as well.

Reply:
> Yes, such kind of processes requires supervision.

Are you referring to background processes like sending patron notices or processes like check in? 

Supervision by who?

> I thought, there should be a centralized logging collection for production (ELK, or similar).

At the moment I don't believe there is any centralised logging for production. I don't think there are folks monitoring the logs of each module.

Reply:
> It guess that most likely if `check-in-storage` will be failing then entire circulation-storage will be failing as well.

I think it is an assumption that the various APIs are all provided by the same module.

If we do make that assumption, I would suggest it is better to fail the whole check in at the earliest opportunity, when there is less likelihood for partial success.

Reply:
> Are you referring to background processes like sending patron notices or processes like check in?

Yes, this kind of processes.

> Supervision by who?

Usually there are some infrastructure monitoring tools that can monitor applications and also can trigger a notification if an error has logged in log file/amount of warnings has reached some threshold.

Let's proceed with synchronous approach then, since we don't have tools/resources to effectively deal with failures for background processes. It will be an improvement point if we're faced with check-in process slow down.

Reply:
> > Are you referring to background processes like sending patron notices or processes like check in?
> 
> Yes, this kind of processes.
> 
> > Supervision by who?
> 
> Usually there are some infrastructure monitoring tools that can monitor applications and also can trigger a notification if an error has logged in log file/amount of warnings has reached some threshold.

I don't know if FOLIO has any infrastructure like that at the moment. I don't think we've established any expectations about folks monitoring the logs.

> Let's proceed with synchronous approach then, since we don't have tools/resources to effectively deal with failures for background processes. It will be an improvement point if we're faced with check-in process slow down.

That would be my preference. I agree that without the tools in place that is safer and that the performance is a concern that will need to be addressed.

Making some parts of this separate asynchronous or background tasks may be part of that. I think it will likely needs deeper changes than that as well, that might involve architectural concerns (e.g. storage modules, every request going via a gateway etc).